ulib/libax/src/cbindings/mod.rs

```diff
@@ -38,27 +43,52 @@ pub unsafe extern "C" fn ax_rand_u32() -> u32 {
     crate::rand::rand_u32()
 }
 
+use spinlock::SpinNoIrq;
+// Lock used by ax_println_str for C apps
+static LOCK: SpinNoIrq<()> = SpinNoIrq::new(());
+
 /// Print a string to the global standard output stream.
 #[no_mangle]
 pub unsafe extern "C" fn ax_print_str(buf: *const c_char, count: usize) -> c_int {
     if buf.is_null() {
         return -axerrno::LinuxError::EFAULT.code();
     }
+
     let bytes = unsafe { core::slice::from_raw_parts(buf as *const u8, count as _) };
     crate::io::stdout().write(bytes).unwrap() as _
 }
 
+/// Printf a string to the global standard output stream. Add a line break.
+#[no_mangle]
+pub unsafe extern "C" fn ax_println_str(buf: *const c_char, count: usize) -> c_int {
+    if buf.is_null() {
+        return -axerrno::LinuxError::EFAULT.code();
+    }
+
+    let _guard = LOCK.lock();
+    let bytes = unsafe { core::slice::from_raw_parts(buf as *const u8, count as _) };
+    let ret = crate::io::stdout().write(bytes).unwrap() as _;
```

Comment:
use `Result::and_then`

Reply:
change to: 
`crate::io::stdout()`
 `    .write(bytes)`
 `    .and_then(|r| crate::io::stdout().write(b"\n").and_then(|_| Ok(r)))`
 `    .unwrap() as _`

Reply:
`.and_then(|| crate::io::stdout().write(b"\n"))`

Reply:
Can use a temporary variable to store `crate::io::stdout()`?

Reply:
Like this? 
`crate::io::stdout()`
 `    .write(bytes)`
 `    .and_then(|r| {`
 `        crate::io::stdout().write(b"\n").unwrap();`
 `        Ok(r)`
 `    })`
 `    .unwrap() as _`

Reply:
Why still unwrap in `and_then`?

Reply:
ok.....then like:
`crate::io::stdout()`
 `    .write(bytes)`
 `    .and_then(|r| {`
 `        crate::io::stdout().write(b"\n");`
 `        Ok(r)`
 `    })`
 `    .unwrap() as _`

Reply:
> .and_then(|| crate::io::stdout().write(b"\n"))

It doest work?

Reply:
> > .and_then(|| crate::io::stdout().write(b"\n"))
> 
> It doesn't work?

Must use `and_then(|_| crate::io::stdout().write(b"\n")`, because the upper line return result<usize>

Reply:
ok

Reply:
Or add a closure to use the `?` expression.

Reply:
use `ax_call_body`? Like:
```
let _guard = LOCK.lock();
ax_call_body!(ax_println_str, {
        let bytes = unsafe { core::slice::from_raw_parts(buf as *const u8, count as _) };
        crate::io::stdout()
            .write(bytes)
            .and_then(|_| crate::io::stdout().write(b"\n"))? as _
    })
```

Reply:
`axbody` will print redundant debug info.

Reply:
change to: 
```
    let _guard = LOCK.lock();
    let bytes = unsafe { core::slice::from_raw_parts(buf as *const u8, count as _) };
    crate::io::stdout()
        .write(bytes)
        .and_then(|_| crate::io::stdout().write(b"\n"))
        .map_or(0, || -LinuxError::EINVAL)
```
Since `puts` needs to return negative value on error, changing into this solves this problem. And `puts` can `return ax_println_str(...)` directly.

Reply:
use `map(|_| 0)`